compute-lab: Add apache2 task (#1)
teodutu/operating-systems: content/chapters/compute/lab/README.md

One such gray area are browsers.
Some open a new tab in a separate process, while others do it in a separate thread instead.

> ## Usage of processes and threads in apache2

Reviewer: ```suggestion
## Usage of Processes and Threads in `apache2`
```

> We'll take a look at how a real-world application - the apache2 http server - makes use of processes and threads.

Reviewer: ```suggestion
We'll take a look at how a real-world application - the `apache2` HTTP server - makes use of processes and threads.
```

> Since the server must be able to handle multiple clients at the same time, it must therefore use some form of concurrency: when a new client arrives, the server offloads the work of interacting with that client to another process or thread.

Reviewer: ```suggestion
Since the server must be able to handle multiple clients at the same time, it must therefore use some form of concurrency.
When a new client arrives, the server offloads the work of interacting with that client to another process or thread.
```

> The choice of whether to use multiple processes or threads is not baked in into the code. Instead, apache2 provides a couple of modules called MPMs (Multi-Processing Modules). Each module implements a different concurrency model and the users can pick whatever module better fits their needs by editing the server configuration files.

Reviewer: ```suggestion
The choice of whether to use multiple processes or threads is not baked into the code.
Instead, `apache2` provides a couple of modules called MPMs (Multi-Processing Modules).
Each module implements a different concurrency model and the users can pick whatever module best fits their needs by editing the server configuration files.
```

The most common MPMs are
- `prefork`: there are multiple worker processes, each process is single-threaded and handles one client request at a time
- `worker`: there are multiple worker processes, each process is multi-threaded, and each thread handles one client request at a time
- `event`: same as `worker` but designed to better handle some particular use cases

In principle, `prefork` provides more stability and backwards compatibility, but it has a bigger overhead. On the other hand, `worker` and `event` are more scalable, and thus able to handle more simultaneous connections, due to the usage of threads. On modern systems `event` is almost always the default.

> ### apache2 live action

Reviewer: ```suggestion
### `apache2` Live Action
```

> Let's run an actual instance of apache2 and see how everything works. Go to `support/apache2` and run `make run`. This will start a container with apache2 running inside.

Reviewer: ```suggestion
Let's run an actual instance of apache2 and see how everything works.
Go to `support/apache2` and run `make run`.
This will start a container with `apache2` running inside.
```
Here and below: stick to one phrase per line.

> Check that the server runs as expected

Reviewer: ```suggestion
Check that the server runs as expected:
```

> ```Bash

Reviewer: ```suggestion
```
```
Specifying `Bash` for code snippets should be limited to bash scripts, not individual commands as that leads to incorrect highlighting. So remove this language hint from all code snippets.

student@os:~$ curl localhost:8080
<html><body><h1>It works!</h1></body></html>
```

Now go inside the container and take a look at running processes:

> ```Bash

Reviewer: ```suggestion
```
```
Ditto.

student@os:~/.../lab/support/apache2$ docker exec -it apache2-test bash

> root@56b9a761d598:/usr/local/apache2# ps -ef

Reviewer: ```suggestion

root@56b9a761d598:/usr/local/apache2# ps -ef
```
Leave one empty line between subsequent commands so they're easier to tell apart.

UID          PID    PPID  C STIME TTY          TIME CMD
root           1       0  0 20:38 pts/0    00:00:00 httpd -DFOREGROUND
www-data       9       1  0 20:38 pts/0    00:00:00 httpd -DFOREGROUND
www-data      10       1  0 20:38 pts/0    00:00:00 httpd -DFOREGROUND
root          25       0  0 20:40 pts/1    00:00:00 bash
root          31      25  0 20:40 pts/1    00:00:00 ps -ef
```

> We see 3 `httpd` processes. The first one, running as root, is the main process, while the other 2 are the workers.

Reviewer: ```suggestion
We see 3 `httpd` processes.
The first one, running as root, is the main process, while the other 2 are the workers.
```

www-data       8       1      17  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       8       1      18  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       8       1      19  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       9       1       9  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       9       1      14  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       9       1      15  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       9       1      20  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       9       1      21  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       9       1      22  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
www-data       9       1      23  0    7 20:56 pts/0    00:00:00 httpd -DFOREGROUND
root          24       0      24  1    1 20:56 pts/1    00:00:00 bash
root          30      24      30  0    1 20:56 pts/1    00:00:00 ps -efL
```

Indeed, each worker has 7 threads. In fact, the number of threads per worker is configurable, as well as the number of initial workers.

> When a new connection arrives, it will be handled by whatever thread is available from any worker. If all the threads are busy, then the server will spawn more worker processes (and therefore more threads), as long as the total number of threads is below some threshold, which is also configurable.

Reviewer: ```suggestion
When a new connection is created, it will be handled by whatever thread is available from any worker.
If all threads are busy, then the server will spawn more worker processes (and therefore more threads), as long as the total number of threads is below some threshold, which is also configurable.
```

Let's see this dynamic scaling in action. We need to create a number of simultaneous connections that is larger than the current number of threads. There is a simple script in `support/apache2/make_conn.py` to do this:

```Bash
student@os:~/.../lab/support/apache2$ python3 make_conn.py localhost 8080
Press ENTER to exit
```

The script has created 100 connections and will keep them open until we press Enter.

Now, in another terminal, let's check the situation inside the container:

```Bash
student@os:~/.../lab/support/apache2$ docker exec -it apache2-test bash

> root@56b9a761d598:/usr/local/apache2# ps -efL

Reviewer: ```suggestion

root@56b9a761d598:/usr/local/apache2# ps -efL
```

www-data     105       1     126  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     105       1     127  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     105       1     130  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     105       1     131  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     105       1     132  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     105       1     133  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     109       1     109  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     109       1     115  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     109       1     116  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     109       1     121  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     109       1     122  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     109       1     123  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
www-data     109       1     124  0    7 21:07 pts/0    00:00:00 httpd -DFOREGROUND
root         146       0     146  0    1 21:10 pts/1    00:00:00 bash

> root         152     146     152  0    1 21:10 pts/1    00:00:00 ps -efL

Reviewer: There are too many lines here. Either trim them (simply delete a few lines in the middle and replace them with a line like this: `[...]`) or, better, lower the thresholds in the configs and the number of connections created by `make_conn.py`. It's very easy to get lost in such a lengthy output.

```

We see a much larger number of threads, as expected.

> ### Practice: investigate apache2 using strace

Reviewer: ```suggestion
### Practice: Investigate `apache2` Using `strace`
```

Use `strace` to discover the server document root. The document root is the path in the filesystem from where httpd serves all the files requested by the clients.

First you will have to stop the running container using `make stop`, then restart it with `make run-privileged`.

Then you will use `strace` inside the container to attach to the worker processes (use the `-p` option for this). You will also have to use `-f` flag with `strace`, so that it will follow all the threads inside the processes.

After you have attached successfully to all worker processes, use the `curl` command to send a request, like the one in the beginning of this section.

Then check the `strace` output to see what files were opened by the server.

> 

Reviewer: You can add a quiz here so students can check if they found the correct document root.

> 

Reviewer: At the end of this section there should be some sort of conclusion that makes some general considerations / suggestions regarding the usage of threads vs processes in `appache2`. I've no idea how that should sound, though.

Author: I was thinking the conclusion just before this section should work, with some adaptations:

```
Therefore, there is also a gray area where there is no clear right answer to this debate.
One such gray area are browsers.
Some open a new tab in a separate process, while others do it in a separate thread instead.
```

Reviewer: Yes, that is fine. Move it here and add s phrase or 2 about web servers and their usage of threads and processes.

> ## Browser Fight

Reviewer: Remove this section, as yours will replace it.
